### ISSUE TITLE
Suggest calling clone on moved values

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#42446]
We should add a `help` message indicating that calling clone where the value was moved is a good idea, but we should only do this when the value *is* cloneable.

```rust
fn main() {
    let a = String::new();
    a;
    a;
}
```

```
error[E0382]: use of moved value: `a`
 --> test.rs:4:5
  |
3 |     a;
  |     - value moved here
4 |     a;
  |     ^ value used here after move
  |
  = note: move occurs because `a` has type `std::string::String`, which does not implement the `Copy` trait

error: aborting due to previous error(s)
```

**Comment by zackmdavis** [comment]
(Caution may be warranted, as [a comment in borrowck suggests](https://github.com/rust-lang/rust/blame/29bce6e22/src/librustc_borrowck/borrowck/mod.rs#L649-L652) that an earlier implementation of this was more trouble than it was worth.)

**Comment by Mark-Simulacrum** [comment]
Hmm, that's true. I see why we don't want this necessarily (Rust has clones everywhere!) but I am also concerned about the other side where users don't know about Clone and nothing tells them that it's a possibility.

**Comment by nikomatsakis** [comment]
I am against this suggestion, because it is far from universal in my experience that `clone()` is the correct solution. I would however think this would be a good topic to cover in the extended error message for E0382 -- i.e., what are the strategies for avoiding moves: borrow, clone, etc.

**Comment by Mark-Simulacrum** [comment]
Yeah, I'm going to close this. It seems that suggesting calling clone isn't the best idea here...

**Comment by Mark-Simulacrum** [comment]
Ah, actually, yeah, covering this in extended documentation is a good idea...

**Comment by estebank** [comment]
### Mentoring instructions

Expand the documentation for [E0382](https://github.com/rust-lang/rust/blob/3df1f7b82d9bede5122ee745cdd4e731abbce892/src/librustc_borrowck/diagnostics.rs#L66-L127) when it talks about `Copy` and `Clone`.

---

@nikomatsakis, @Mark-Simulacrum, would it make sense to check for `Clone` and add a note, not a suggestion?

      = note: move occurs because `a` has type `std::string::String`, which does not implement the `Copy` trait
      = help: `std::string::String` implements the `Clone` trait, so you could call `.clone()` on `a` to copy its contents to a new location in memory

**Comment by Mark-Simulacrum** [comment]
I don't think adding the help would be a good idea. I share @nikomatsakis' concerns over the potential negative impacts, along the lines of "why no auto clone, then?" which would probably be non-ideal for us. I think the extended error description is the perfect place to put this, since newcomers to Rust hopefully are using that as much as possible (do we have any data on this?).

**Comment by estebank** [comment]
> I don't think adding the help would be a good idea.

Fair enough. We can close this ticket then when we expand the error docs.

> do we have any data on this?

I don't know if we do, as I haven't seen any code in the compiler to send metrics (but I haven't looked for that either).

Now that I think about it, it might be a neat feature to highlight very prominently the `--explain` flag when first running `rustc` and always highlight it when doing `rustc -h`, or at the very least make it the first line of the help (instead of now somewhere in the lower half).

**Comment by CAD97** [comment]
Is this fixed by #45082 or do we want more?
